### PR TITLE
Inject mock annotations cleanup

### PR DESCRIPTION
**Pull request by mkouba**
- introduce io.quarkus.test.InjectMock
- deprecate io.quakus.test.junit.mockito.InjectMock
- remove io.quarkus.test.component.ConfigureMock

### CODE REVIEW

File: test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/VerifyMockitoMocksCallback.java
Comment (mkouba) [inline test-framework/junit5/src/main/java/io/quarkus/test/junit/internal/VerifyMockitoMocksCallback.java:28]: We need this because the `@io.quarkus.test.InjectMock` annotation is on the class path even if not supported (i.e. no `quarkus-junit5-mockito` dependency)...

**Comment by geoand** [comment]
I have not been following the latest discussions because I was out for a few days, so I'll let others review before I do

**Comment by glefloch** [comment]
This looks cool to have a common annotation for mocks and the ability to provide different implementation. I'm maintaining this project which make it possible to use `Mockk` as mocking library (usually used in kotlin). With this, we could rely on the same `InjectMock` annotation. 
There is also an `@InjectSpy` annotation, will it be part of the common package too ?

**Comment by mkouba** [comment]
> There is also an `@InjectSpy` annotation, will it be part of the common package too ?

That is a very good question. I don't know. 

For example, `QuarkusComponentTest` does not support mockito spies. 

And given the fact that `Mockito#spy()` javadoc states: _"Real spies should be used <b>carefully and occasionally</b>, for example when dealing with legacy code."_... :shrug:

File: test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java
Comment (Ladicek) [inline test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java:11]: Maybe mention that this only makes sense for `@QuarkusTest` and is useless/ignored for `@QuarkusComponentTest`?

Comment (mkouba) [reply test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java:11]: Probably makes sense...

Comment (holly-cummins) [reply test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java:11]: So this is totally about my understanding, not about the code, and not about this particular section of code, but.
 
 My impression - based totally on @mkouba's demo, not on actually reading the code - was that if you used `@QuarkusComponentTest`, you were setting the system under test to be narrow, and therefore other parts of the system were replaced with mocks, and that we'd default to using Mockito for those mocks. 

I have to admit I was always a bit hazy about where the mockito came from, if people didn't have the `junit-5-mockito` dependency on their classpath ... but if my broad understanding is correct, don't we need mechanisms to adjust the behaviour of mockito in the `@QuarkusComponentTest` scenario?

File: docs/src/main/asciidoc/getting-started-testing.adoc
Comment (holly-cummins) [inline docs/src/main/asciidoc/getting-started-testing.adoc:837]: Do people also need to have the `quarkus-junit5-mockito` dependency in their project, if they want mockito to be doing the mocking?

Comment (holly-cummins) [reply docs/src/main/asciidoc/getting-started-testing.adoc:837]: Oh, and for the linter comment, IANATW, but I'd suggest 'with', rather than 'via' or 'through'. 

"This functionality is available with the `@io.quarkus.test.InjectMock` annotation."

I'm not totally happy with 'with', though, so happy to be over-ruled.

File: docs/src/main/asciidoc/getting-started-testing.adoc
Comment (holly-cummins) [inline docs/src/main/asciidoc/getting-started-testing.adoc:831]: I suggest 

"So if you call methods such as `verify` you should hang on to the mock instance in your test, or use `@io.quarkus.test.InjectMock`"

File: test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java
Comment (mkouba) [reply test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java:11]: > I have to admit I was always a bit hazy about where the mockito came from, if people didn't have the junit-5-mockito dependency on their classpath...

`quarkus-junit5-component` depends on mockito as well... no `quarkus-junit5-mockito` needed.

**Comment by holly-cummins** [comment]
I think this looks great. 

Adding the back-link to https://github.com/quarkusio/quarkus/issues/33827. Once this merges I'll tidy up my table there.

File: docs/src/main/asciidoc/getting-started-testing.adoc
Comment (mkouba) [reply docs/src/main/asciidoc/getting-started-testing.adoc:837]: Yes, the `quarkus-junit5-mockito` dependency is still needed. So what about:

"This functionality is available with the `@io.quarkus.test.InjectMock` annotation if the `quarkus-junit5-mockito` dependency is present."

File: test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java
Comment (holly-cummins) [reply test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java:11]: Ah, got it. So if they wanted to use `@QuarkusComponentTest` and another mocking framework, they'd be 
- out of luck for the actual `@QuarkusComponentTest`s, which is reasonable
- maybe out of luck elsewhere if they try and use the new `@InjectMock, but since we haven't really written the pluggable back-end for that, maybe not?

File: docs/src/main/asciidoc/getting-started-testing.adoc
Comment (holly-cummins) [reply docs/src/main/asciidoc/getting-started-testing.adoc:837]: If they have `quarkus-junit5-component` and not `quarkus-junit-mockito`, is the behaviour that `@InjectMock` would work in a `@QuarkusComponentTest` (since there's actually a `@ConfigureMock` in disguise :) ), but not work in other tests? 

That could be confusing, but I have no idea what could be done to make it less confusing.

**Review by holly-cummins — APPROVED** [review]

File: test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java
Comment (mkouba) [reply test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/MockitoConfig.java]: Done.

**Comment by mkouba** [comment]
FTR I've created https://github.com/quarkusio/quarkus/issues/34018 for a common `@InjectSpy`.

**Review by Ladicek — APPROVED** [review]

File: test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SingletonToApplicationScopedTestBuildChainCustomizerProducer.java
Comment (luneo7) [inline test-framework/junit5-mockito/src/main/java/io/quarkus/test/junit/mockito/internal/SingletonToApplicationScopedTestBuildChainCustomizerProducer.java:57]: I think there was a typo here maybe this should have been:
```
if (instance.name().equals(MOCKITO_CONFIG) && instance.target().asField().hasAnnotation(DEPRECATED_INJECT_MOCK)) {
   continue;
}
```
As I think that we shouldn't process `@MockConfig` and the deprecated  `@InjectMock` twice as it sets the same configs, but we should process the `@MockConfig` if being used in conjunction with the new `@InjectMock` to tweak the mock configuration.
Since now using `@MockitoConfig` to convert scopes doesn't work =/
Also saw that there is no test for converting scopes, so this it wasn't caught before.
Good thing is that there is still the deprecated `@InjectMock` as we can use it do to scope conversion.

**Pull request merged**
